src/kubernetes/api_watcher.rs

```diff
@@ -152,3 +164,457 @@ pub mod invocation {
+                    when.method(GET).path("/api/v1/pods");
+                    then.status(410)
+                        .header("Content-Type", "application/json")
+                        .body("body");
+                }),
+                Some(StatusCode::GONE),
+                true,
+            ),
+            // Other error.
+            (
+                Box::new(|when, then| {
+                    when.method(GET).path("/api/v1/pods");
+                    then.status(400)
+                        .header("Content-Type", "application/json")
+                        .body("body");
+                }),
+                Some(StatusCode::BAD_REQUEST),
+                false,
+            ),
+        ];
+
+        for (mock_config, expected_bad_status, expected_is_desync) in cases {
+            let server = MockServer::start_async().await;
+            let mock = server.mock_async(mock_config).await;
+
+            let config = client::Config {
+                base: server.base_url().parse().unwrap(),
+                token: "SOMEGARBAGETOKEN".to_string(),
+                tls_options: TlsOptions::default(),
+            };
+            let client = Client::new(config).unwrap();
+            let mut api_watcher = ApiWatcher::new(client, Pod::watch_pod_for_all_namespaces);
+            let error = api_watcher
+                .watch(WatchOptional {
+                    allow_watch_bookmarks: Some(true),
+                    field_selector: None,
+                    label_selector: None,
+                    resource_version: Some(""),
+                    timeout_seconds: Some(300),
+                    pretty: None,
+                })
+                .await
+                .err()
+                .expect("expected an invocation error here");
+
+            let (actual_status, actual_is_desync) = match error {
+                watcher::invocation::Error::Desync {
+                    source: invocation::Error::BadStatus { status },
+                } => (Some(status), true),
+                watcher::invocation::Error::Other {
+                    source: invocation::Error::BadStatus { status },
+                } => (Some(status), false),
+                watcher::invocation::Error::Desync { .. } => (None, true),
+                watcher::invocation::Error::Other { .. } => (None, false),
+            };
+
+            assert_eq!(
+                actual_status, expected_bad_status,
+                "actual left, expected right"
+            );
+            assert_eq!(
+                actual_is_desync, expected_is_desync,
+                "actual left, expected right"
+            );
+            mock.assert_async().await;
+        }
+    }
+
+    /// Test that it can handle stream errors.
+    #[tokio::test]
+    async fn test_stream_errors() {
+        let cases: Vec<(
+            Box<dyn FnOnce(When, Then)>,
+            Vec<Box<dyn FnOnce(Result<WatchResponse<Pod>, watcher::stream::Error<stream::Error>>)>>,
+        )> = vec![
```

Comment:
I think we want to be able to trigger and test more of the possible stream item variants.

What's already covered:
- `Ok(WatchResponse::Ok(WatchEvent::Added { ... }))`
- `Err(watcher::stream::Error::Desync { ... }))` - should assert the underlying value (the `...` should be a concrete value matching the test input to assert the data is porperly passed)
- `Err(watcher::stream::Error::Other { ... }))` - same as above
- `Ok(WatchResponse::Ok(WatchEvent::ErrorStatus(Status { ... })))` (covered by the non-desync status error case)

What we probably should cover as well:
- `Ok(WatchResponse::Ok(WatchEvent::ErrorOther(RawExtension { ... })))` - pass the response event with some non-standard `type` (i.e. `extension_test`)
- `Ok(WatchResponse::Other(Ok(Some(serde_json::Value { ... })))` - pass a valid JSON that is not a valid `WatchResponse` - sth like `{"a":"b"}` should do
- `Ok(WatchResponse::Other(Ok(None))` - pass nothing?
- `Ok(WatchResponse::Other(Err(...))` - I presume this should be the value when passing an invalid JSON? It would be great to add a test that illustrates how to trigger this error